packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js

```diff
@@ -0,0 +1,329 @@
+      let setValue = null;
+      function App() {
+        let [value, _setValue] = React.useState(0);
+        setValue = _setValue;
+        return (
+          <button id="button" onClick={() => setValue(2)}>
+            {value}
+          </button>
+        );
+      }
+      const container = document.createElement('div');
+      document.body.appendChild(container);
+      let button;
+      act(() => {
+        ReactDOM.render(<App />, container);
+        button = container.querySelector('#button');
+        button.dispatchEvent(new MouseEvent('click', {bubbles: true}));
+      });
+      expect(button.innerHTML).toBe('2');
+      expect(() => setValue(1)).toWarnDev([
+        'An update to App inside a test was not wrapped in act(...).',
+      ]);
+      document.body.removeChild(container);
+    });
+    describe('fake timers', () => {
+      beforeEach(() => {
+        jest.useFakeTimers();
+      });
+      afterEach(() => {
+        jest.useRealTimers();
+      });
+      it('lets a ticker update', () => {
+        function App() {
+          let [toggle, setToggle] = React.useState(0);
+          React.useEffect(() => {
+            let timeout = setTimeout(() => {
+              setToggle(1);
+            }, 200);
+            return () => clearTimeout(timeout);
+          }, []);
+          return toggle;
+        }
+        const container = document.createElement('div');
+
+        act(() => {
+          ReactDOM.render(<App />, container);
+        });
+        act(() => {
+          jest.runAllTimers();
+        });
+
+        expect(container.innerHTML).toBe('1');
+      });
+    });
+
+    it('warns if you return a value inside act', () => {
+      expect(() => act(() => null)).toWarnDev(
+        [
+          'The callback passed to act(...) function must return undefined, or a Promise.',
+        ],
+        {withoutStack: true},
+      );
+      expect(() => act(() => 123)).toWarnDev(
+        [
+          'The callback passed to act(...) function must return undefined, or a Promise.',
+        ],
+        {withoutStack: true},
+      );
+    });
+
+    it('warns if you try to await an .act call', () => {
+      expect(act(() => {}).then).toWarnDev(
+        [
+          'Do not await the result of calling act(...) with sync logic, it is not a Promise.',
+        ],
+        {withoutStack: true},
+      );
+    });
+  });
+  describe('asynchronous tests', () => {
+    it('can handle timers', async () => {
+      function App() {
+        let [ctr, setCtr] = React.useState(0);
+        function doSomething() {
+          setTimeout(() => {
+            setCtr(1);
+          }, 200);
+        }
+
+        React.useEffect(() => {
+          doSomething();
+        }, []);
+        return ctr;
+      }
+      const el = document.createElement('div');
+      await act(async () => {
+        act(() => {
+          ReactDOM.render(<App />, el);
+        });
+
+        await sleep(500);
```

Comment:
So the reason I don't like using native timers in all of these tests is that now it takes at least 500ms before this test can finish. Whereas if you use fake timers, it's CPU bound.

Reply:
I can reduce the timer length, but this is specifically to test when we can't use fake timers reliably.

Reply:
It's fine to use real timers in one or two of these tests but doing it in all of them seems like overkill especially since most people who use timers are going to use the fake timers, anyway.